Add port usage (#3395)
istio/istio.io: content/docs/setup/kubernetes/spec-requirements/index.md

* _**`NET_ADMIN` capability**:_ If pod security policies are enforced in your cluster and unless you use the [Istio CNI Plugin](/docs/setup/kubernetes/istio-cni-install/), your pods must have the `NET_ADMIN` capability allowed.
See [Required Pod Capabilities](/help/ops/setup/required-pod-capabilities/).

> * _**Service Ports and Protocols**_: Ensure that your service's ports and protocols are available and don't conflict with the [Istio used ports and protocols](#istio-used-ports) for the Istio components be used.

Reviewer: ```suggestion
* _**Service Ports and Protocols**_: Ensure your service's ports and protocols are available and don't conflict with the [ports and protocols Istio uses](#istio-used-ports) for its components.
```

Author: Removed section per @rshriram comments

> 

Reviewer 2: Please remove this sentence. There is no conflict with any Istio ports on Kubernetes. ever.

Author: Done

> ## Istio Used Ports

Reviewer 3: We don't capitalize headings.

How about "Istio port usage" instead?

Reviewer: This is one of the few instances where I would use passive voice to avoid a weird sentence and to emphasize "Ports":
```suggestion
## Ports used by Istio
```

Author: Done (along with references)

> The following ports and protocols are used by Istio.

Author: Done.

> If Istio is using HTTP on a port, a service may also use HTTP, but not TCP on that port. This is because HTTP routing is dependent on a number of factors such as Hostname, headers or path which removes the conflict.

Author: Removed section per @rshriram comments

> If Istio is using TCP, another service may not use HTTP on that port. The service may potentially use TCP on that port, marked with a `?` in the table, by using SNI headers or IP address blocks. For example, see [Consuming External TCP Services](/blog/2018/egress-tcp/).

Reviewer: ```suggestion
If Istio uses a port for TCP traffic, another service usually **can't** use HTTP on that port. Using SNI headers or IP address blocks, the service could potentially use that port for TCP. If that is the case, the port is marked with an interrogation sign, `?` in the table below. For an example, see [Consuming External TCP Services](/blog/2018/egress-tcp/).
```

Author: Removed section per @rshriram comments

> 

Reviewer 2: neither of the above two are true. I understand the Kafka issue but the warnings here do nothing to address that.
Simply state that do not have any TCP headless services on the same port as Istio's TCP ports. Thats it

Author: Done.

| 9901 | HTTP | Yes | No |  Galley | |
| 14267 | TCP | No | `?` |  Tracing  |  Tracing - Jaeger |
| 14268 | TCP | No | `?` |  Tracing  |  Tracing - Jaeger |
| 15000 | TCP | No | `?` |  Envoy  |  Envoy admin port (commands/diagnostics) |
| 15001 | TCP | No | `?` |  Envoy  |  Envoy |
| 15004 | HTTP | Yes | `?` |  Mixer, Pilot |  Policy/Telemetry - `mTLS` |
| 15010 | HTTP | Yes | No |  Pilot  |  Pilot service - XDS pilot - discovery |
| 15011 | TCP | No | `?` |  Pilot  |  Pilot service - `mTLS` - Proxy - discovery |
| 15014 | HTTP | Yes | No |  Citadel, Galley, Mixer, Pilot  |  Control plane monitoring |
| 15029 | TCP | No | `?` |  Ingress gateway , Kiali |  Kiali |
| 15030 | TCP | No | `?` |  Ingress gateway, Prometheus |  Prometheus |
| 15031 | TCP | No | `?` |  Grafana, Ingress Gateway  |  Grafana |
| 15032 | TCP | No | `?` |  Ingress gateway, Tracing  |  Tracing |
| 15090 | HTTP | Yes | No |  Egress gateway, Ingress Gateway, Mixer  |  Proxy |
| 15443 | TCP | No | `?` |  Egress gateway, Ingress Gateway  |  This is the port where SNI routing happens |
| 16686 | TCP | No | ? |  Tracing  |  Tracing - Jaeger |
| 20001 | HTTP | Yes | No |  Kiali | |

> | 42422 | TCP | No | ? |  Mixer  |  Telemetry - Prometheus |

Reviewer 4: How often does this change? How will we keep it up to date?

Author: We will need to automate this as well. Probably need a table of port descriptions, use `istioctl` to get the list of ports, and then build table from that. Since we are limiting the ports to certain services, a blacklist of those used by the other services would be good. Any ports that don't have a description and aren't in the blacklist would fail to run and require some intervention to get a new port description.

> | 42422 | TCP | No | ? |  Mixer  |  Telemetry - Prometheus |

Reviewer 2: this giant list is very scary. Not to mention most of these are unused in production when I don't install addons like kiali or grafana. 

Please trim this down to the ports used by mixer, pilot and citadel.

Author: I remove a number of ports. I did leave Prometheus and Envoy ports in the chart since they are in the production profile.

    | | | | |
    | Galley | 443 |   | Validation port |
    | Galley | 9091 |   | Mesh control Plane - Running mixer |
    | Galley | 9094 |   | Port to use for exposing profiling (no longer used?) |
    | Galley | 15014 |   | Monitoring |
    | | | | |
    | Grafana | 3000 |   | Grafana |
    | Grafana | 15031 |   | Grafana |
    | | | | |
    | Envoy | 15000 |   | Envoy admin port (commands/diagnostics) |
    | | | | |
    | Ingress gateway | 80 | 31380 |   |
    | Ingress gateway | 443 | 31390 |   |
    | Ingress gateway | 15029 |   | Kiali |
    | Ingress gateway | 15030 |   | Prometheus |
    | Ingress gateway | 15031 |   | Grafana |
    | Ingress gateway | 15032 |   | Tracing |
    | Ingress gateway | 15090 |   | Proxy |
    | Ingress gateway | 15443 |   | This is the port where SNI routing happens |
    | Ingress gateway | 31400 | 31400 | Sample port|
    | | | | |
    | Mixer | 9091 |   | Policy/Telemetry |
    | Mixer | 15004 |   | Policy/Telemetry - `mTLS`|
    | Mixer | 15014 |   | Policy/Telemetry monitoring port |
    | Mixer | 15090 |   | Proxy |
    | Mixer | 42422 |   | Telemetry - Prometheus |

>     | | | | |

Reviewer 5: could we separate default on component from default off component?

Could we use policy and telemetry as that is what user sees in k get services or deploy?

Author: I used Mixer since it's what the chart is named (that column is by chart name). I could change _Mixer_ to _Policy, Telemetry_ in to better coincide with the pod names if others feel that makes more sense.

    | Pilot | 8080 |   | Pilot service - Discovery - legacy |

>     | Pilot | 15003 |   | Pilot service - Proxy - http - Discovery - old |

Reviewer 5: do we still use this?  I don't see it used in minimal profile I'm running.

Author: I don't either, but it was in the original yaml. Removed in latest chart.

    | Pilot | 15004 |   | Pilot service - Mixer |
    | Pilot | 15005 |   | Pilot service - Proxy - https - Discovery |

>     | Pilot | 15007 |   | Pilot service - Proxy - http - Discovery |

Reviewer 5: is 15005 or 15007 used?  I don't see 
I don't see it used in minimal profile I'm running.

Author: Removed in latest chart

    | Pilot | 15010 |   | Pilot service - XDS pilot - discovery |
    | Pilot | 15011 |   | Pilot service - `mTLS` - Proxy - discovery |

>     | Pilot | 15012 |   | Pilot service - `mTLS` - Proxy - Discovery service grpc address, with https discovery (still used?) |

Reviewer 5: seems not used either?

Author: removed as well.

    | Pilot | 15014 |   | Pilot service - Monitoring |
    | | | | |
    | Prometheus | 9090 |   | Prometheus |
    | | | | |

>     | Service Graph | 8088 |   | Service Graph |

Reviewer 5: isn't service graph deprecated in 1.1?

Author: I still saw it running when deploying the demo profile.

Author: Removed.

    | | | | |
    | Sidecar injector | 443 |   | Sidecar injector |
    | | | | |
    | Tracing | 80 |   | Tracing - Jaeger |
    | Tracing | 9411 |   | Tracing - Other |
    | Tracing | 15032 |   | Tracing - Jaeger |
    | Tracing | 16686 |   | Tracing - Jaeger |
    | | | | |

>     | Zipkin | 9411 |   | Zipkin collector |

Reviewer 5: Ok, I feel we also need to separate the ports to ports that user should care when moving their services to Istio which runs on the proxy mostly, and ports admin should care.  WDYT?

Reviewer 4: I agree. This big table is not useful. There should just be a list entry in this document for the ports that pods can't use because of istio-proxy port conflict. As I commented earlier, if there is a table needed to list the purpose of these ports, it should be in a separate section of the document, and just referenced from the pod requirements list entry.

Other component ports that need to be free to deploy Istio should be documented separately and listed in a different table. Which ports have this problem and why? Clearly not all of them in this big list.

Author: I'll talk with @GregHanson again, but I think we have two sets of ports/protocols that services can't use: 1) Ports because they are already in use in the `Pod`, 2) Ports/Protocols because Pilot fails to push rules if there is a conflict. The later is a larger list and what this table describes. At least that is how I understand the problem and trying to document.

* _**`NET_ADMIN` capability**:_ If pod security policies are enforced in your cluster and unless you use the [Istio CNI Plugin](/docs/setup/kubernetes/istio-cni-install/), your pods must have the `NET_ADMIN` capability allowed.
See [Required Pod Capabilities](/help/ops/setup/required-pod-capabilities/).

> * _**Check required ports**_:

Reviewer 4: How is this an application pod/service requirement? Check ports how?

Author: Added some text on verifying ports. Basically, one has to verify that the Istio ports are available before installing Istio or there will be a port conflict. Similarly, their services may not start post install if they are using a port that Istio is already using. These would be requirements on their services so why I thought to put on this page. I'm fine moving it to a better spot.

Reviewer 4: First of all, I don't think it makes any sense to embed this big table in the bulleted list of pod/service requirements. If you want to put the table in the same document, at least move it to a separate section, for example:
```
## Istio Required Ports

Make sure the following ports are available before installing Istio components.

<table> ...
```

Then, you can have a entry in the pod/service requirements list that says something like:
```
***Pod and Service Ports***: Ensure that your ports are available and don't conflict with the [Istio required ports](#istio_required_ports).
```
That said, I'm not sure exactly what the actual requirement is. Is it pod ports can't conflict, service ports can't conflict, both? HTTP ports, TLS ports, both? Does the NodePort also matter? You need to document the restriction accurately.

Author: I did put the table in a different section and referenced from the bulled list.

I also restructured the table to list by port and include the list of components using the port/protocol.

Greg also helped with creating some text to describe the issues between protocol usage for each port.

>     | Component | Port | NodePort | Description |

Reviewer: Where did the table come from? How will it be updated? Why are you listing all ports instead of just the ones that can be overridden?

Author: Table was generated by me from several places: issues, yaml files, source code.

It would be up to developers to update the table when ports are changed/added.

Istio users would need to make sure all the ports for any services they install are available

I expect as work continues we will allow more ports to be overridable and some to move to a range specific to Istio.

Reviewer 5: thoughts on making this table its own page and link to it from the pod requirement list here?   Seems this is something deserves its own page.

Author: The recent incantation of the table was made using `istioctl proxy-config listeners` and merging with data I already had. I also used `istioctl proxy-config enpoint|clusters`.

    | Mixer | 15090 |   | Proxy |
    | Mixer | 42422 |   | Telemetry - Prometheus |
    | | | | |
    | Pilot | 8080 |   | Pilot service - Discovery - legacy |
    | Pilot | 15003 |   | Pilot service - Proxy - http - Discovery - old |
    | Pilot | 15004 |   | Pilot service - Mixer |
    | Pilot | 15005 |   | Pilot service - Proxy - https - Discovery |
    | Pilot | 15007 |   | Pilot service - Proxy - http - Discovery |
    | Pilot | 15010 |   | Pilot service - XDS pilot - discovery |
    | Pilot | 15011 |   | Pilot service - `mTLS` - Proxy - discovery |
    | Pilot | 15012 |   | Pilot service - `mTLS` - Proxy - Discovery service grpc address, with https discovery (still used?) |
    | Pilot | 15014* |   | Pilot service - Monitoring |
    | | | | |
    | Prometheus | 9090 |   | Prometheus |
    | | | | |
    | Service Graph | 8088 |   | Service Graph |
    | | | | |
    | Sidecar injector | 443 |   | Sidecar injector |
    | | | | |
    | Tracing | 80 |   | Tracing - Jaeger |
    | Tracing | 9411 |   | Tracing - Other |
    | Tracing | 15032 |   | Tracing - Jaeger |
    | Tracing | 16686 |   | Tracing - Jaeger |
    | | | | |
    | Zipkin | 9411 |   | Zipkin collector |

> Any port numbers marked with * are overridable, so you will need to ensure any

Reviewer: ```suggestion
You can override any port numbers marked with an asterisk `*`.  Ensure any
```

Author: Changed and moved text to before the paragraph.
